src/main/java/org/javarosa/core/io/StreamsUtil.java

```diff
@@ -118,33 +120,54 @@ public static void writeFromInputToOutputUnmanaged(InputStream is, OutputStream
     /**
      * Write is to os and close both
      */
-    public static void writeFromInputToOutputNew(InputStream is, OutputStream os) throws IOException {
-        writeFromInputToOutputNew(is, os, null);
+    public static void writeFromInputToOutputNew(InputStream is, OutputStream os) throws InputIOException, OutputIOException {
+        writeFromInputToOutputNewInner(is, os, null);
     }
 
     /**
      * Write is to os and close both
      */
-    public static void writeFromInputToOutputNew(InputStream is, OutputStream os, StreamReadObserver observer) throws IOException {
+    public static void writeFromInputToOutputNew(InputStream is, OutputStream os, StreamReadObserver observer) throws InputIOException, OutputIOException {
+        writeFromInputToOutputNewInner(is, os, observer);
+    }
+
+    /**
+     * Write is to os and close both
+     */
+    private static void writeFromInputToOutputNewInner(InputStream is, OutputStream os, @Nullable StreamReadObserver observer)
+            throws InputIOException, OutputIOException {
         byte[] buffer = new byte[8192];
         long counter = 0;
 
         try {
-            int count = is.read(buffer);
+            int count = readIntoBuffer(is, buffer);
             while (count != -1) {
                 counter += count;
                 if (observer != null) {
                     observer.notifyCurrentCount(counter);
                 }
-                os.write(buffer, 0, count);
-                count = is.read(buffer);
+
+                try {
+                    os.write(buffer, 0, count);
```

Comment:
Nit: this line and the line on 143 above have the same structure, but with different patterns to handle them (one externalized, one internalized) which seems odd to me.